### ISSUE TITLE
Diskmanipulator import needs overwrite option

### ISSUE THREAD

**Issue by D15C0DE** [openMSX/openMSX#1393]
As mentioned on IRC I'm trying to use a directory on my Ubuntu PC as a large dir-as-dsk using a 32MB harddisk image and diskmanipulator import/export.

I use this script:
```
set power off
ext ide
hda test.img
diskmanipulator import hda test
set power on
bind CTRL+PAUSE "diskmanipulator export hda testoutput"
```

I use a different directory for output because export doesn't preserve the filestamps and there are filename case-sensitivity issues.

However, this issue report is about the problem that some files which are modified in the test directory are not being updated in the image file during the import. Some are, some aren't. There doesn't seem to be a logic to it.

**Comment by D15C0DE** [comment]
Also note that the Diskmanipulator manual has various errors, such as examples that don't work.

**Comment by m9710797** [comment]
[Lunch-break@work, I couldn't do any experiments yet, nor did I look at any source code].

I agree the above scenario should either just work or produce a clean error message. The following questions/remarks are only to help narrow it down.
- I've used a similar script myself in the past. Though the main difference was that I always started from a clean image (with "diskmanipulator create") while you continue from an existing image.
- Could it be that the image is (nearly) full? (Even if the host-directory "test" contains less than 32MB data, it might be possible there are more leftover files in the image).
- Could you run the MSXDOS CHKDSK tool on the image? Possibly the image got corrupted in some way (in the past) (either via diskmanipulator or via some other way).
- Are there any subdirectories in use? Do the corrupt files appear (only) inside subdirs or (also) in the root dir? (Just to narrow down the bug-search-space).

**Comment by D15C0DE** [comment]
I don't use subdirectories. There are 189 files on the image, in total less than 5MB. As far as I know IDE supports up to 512 files in the root directory. All files are accessible normally. None are corrupt. The only problem is that "diskmanipulator import" does not replace some of the files with the latest versions, even if it does replace others. The files in question have no case-sensitivity issues. If I use MSX-DOS2 to delete those files, running "diskmanipulator import" again will put the newer files on the image.

**Comment by m9710797** [comment]
Hi, I found some time to investigate this issue. I believe I found the cause (which may or may not be a satisfying solution for you).

Currently the 'diskmanipulator import' command will _never_ update existing files in the disk image. It will only import _new_ files. In your first message you mention that some files are and some aren't updated. Could you double check? All the 'updated' files should really be new files.

When you run the 'diskmanipulator import' command, it will even print warnings for all such files. For example:
`    Warning: preserving entry myfile.txt`
Though because you ran this command from a script (and didn't do anything with the output of the command), you couldn't see this.

We could extend the 'diskmanipulator import' command with an option to overwrite existing files. (But should this also delete whole subdirs in the diskimage if a hostfile with the same name gets imported? Might be dangerous). Shorter term, as a workaround, you could instead of importing in an existing image import into a clean image. Or import into a (new) subdirectory in the existing image.

**Comment by D15C0DE** [comment]
I regenerate 4 files just before starting openMSX, one of them always has its most recent timestamp while the other 3 don't change. I checked whether the file contents are updated as well, and as you suggest they are not. So that is a bug. The timestamp in the image should not be changed if the file is not overwritten.

I ran the diskmanipulator import command in the console and indeed saw those warnings, even for the file that got its filestamp updated.

Creating a clean image every time means having to manually manage files I generate from within the MSX with files I generate from the PC side. It is not ideal, even more so because I can't bind diskmanipulator export to ALT+F4 (it quits the emulator without executing the binded command) and there is no "after exit" command either. So if I forget to export the files I need manually, they are lost when I recreate the image file.

**Comment by m9710797** [comment]
That's indeed very strange. I can't explain it.

I'll try to give an overview of the current implementation. Then maybe other people can help and point me to what I'm overlooking. (I'm skipping stuff related to sub-directories).
- For a host-directory the 'diskmanipulator import' command calls MSXtar::addDir() (in src/fdc/Diskmanipulator.cc on line 506).
- addDir() (src/fdc/MSXtar.cc:847), delegates to MSXtar::recurseDirFill() (src/fdc/MSXtar.cc:617).
- recurseDirFill() loops over the entries in the host-directory. For each file it calls addFileToDSK() (src/fdc/MSXtar.cc:623).
- addFileToDSK() takes the host-filename and transforms it into a 8.3 msx-filename (src/fdc/MSXtar.cc:581).
- Then it tries to find this msx-filename in the current disk-image (src/fdc/MSXtar.cc:585).
- Only if there was no such msx-file yet do we create it (src/fdc/MSXtar.cc:592), set the timestamp (src/fdc/MSXtar.cc:602) and import the file-content (src/fdc/MSXtar.cc:605).

In other words: I don't see how the current code could make any changes (timestamp and/or content) to existing msx-files. What am I missing?

**Comment by D15C0DE** [comment]
Thanks for that analysis! I'm sorry. It made me realize that the file in question is rewritten in the autoexec.bat. So there is no bug.

I see // TODO implement overwrite option
Maybe I will look into adding that myself.

**Comment by D15C0DE** [comment]
I fixed the issue with the timestamps on diskmanipulator export.
diskmanipulator export always overwrites files, so in my opinion import should do the same.

**Comment by D15C0DE** [comment]
Which I have implemented now. I can now safely import (at startup) and export (at exit) to the same directory, without losing timestamps and with the ability to manually add files from either the MSX or the PC side at any time.

Compared to dirasdsk it just doesn't handle file removals, filename case-sensitivity, or automatic syncing.

**Comment by m9710797** [comment]
Ok, I'm glad it's resolved. I'm closing this issue now.
Feel free to still send a PR with your changes.
